Ensemble-Model.md

```diff
@@ -0,0 +1,1066 @@
+
+In the following, we perform these operations:
+- [Min-Max scaling](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html)
+- Changing the locale scores to a double square root
+
+
+```python
+max_feature_values = X_unnormalized.map(lambda (_, features): np.max(features, axis=0)).reduce(np.maximum)
+```
+
+
+```python
+def preprocess_locale_scores(scores):
+    return np.sqrt(np.sqrt(scores))
+```
+
+
+```python
+def scale_features((client, features)):
+    features = features / max_feature_values
+    features[:, 0] = preprocess_locale_scores(features[:, 0])
+    return client, features
+```
+
+
+```python
+X = X_unnormalized.map(scale_features).cache()
+```
+
+## Making recommendations
+
+Computing recommendations then reduces down to a dot product. These results are then sorted.
+
+
+```python
+def get_weighted_recommendations(client_data, features, weights):
+    scores = features.dot(weights)
+    return client_data, np.argsort(-scores)
+```
+
+## Measuring the performance (MAP)
+
+We use the [MAP](https://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html) measure as an error metric for this optimization problem. The reason for this is mostly that we only have positive data, i.e. we know addons which users like, but we don't really have a lot of data about addons that users hate.
+
+
+```python
+def average_precision(client_data, recommendations):
+    tp = fp = 0.
+    masked = set(client_data['masked_addons'])
+    precisions = []
+    
+    for recommendation in recommendations:
+        if whitelist[recommendation] in masked:
+            tp += 1
+            precisions.append(tp / (tp + fp))
+            if tp == len(masked):
+                break
+        else:
+            fp += 1
+    
+    if len(precisions) > 0:
+        return np.mean(precisions)
+    else:
+        return 0.
+```
+
+## Training an ensemble model
+
+
+```python
+from scipy.optimize import minimize
+```
+
+### Defining a cost function
+
+We find recommendations, compute the average precision (AP) and then calculate the mean of that (MAP). This produces a value between 0 and 1. We then subtract this value from 1 because SciPy looks for a function to minimize, not to maximize.
+
+
+```python
+def cost(weights):
+    return 1 - X.map(lambda (client_data, features): get_weighted_recommendations(client_data, features, weights))\
+            .map(lambda (client_data, recommendations): average_precision(client_data, recommendations))\
+            .mean()
+```
+
+### Choosing an initial guess
+
+There are many ways of choosing initial guesses. A constant vector of 1s seems to be a sensible prior (with properly normalized features it means that all recommenders are equally useful). However, randomly choosing initial values can also be useful.
+
+
+```python
+def get_initial_guess(n):
```

Comment:
Please leave one strategy and disable/comment the other. When running this notebook, the random initialization will always prevail otherwise. Let's make the choice explicit.